### ISSUE TITLE
distributed RPC-based speculative evaluation

### ISSUE THREAD

**Issue by okuvshynov** [okuvshynov/llama_duo#1]
plan copypasta from https://github.com/ggerganov/llama.cpp/discussions/6853#discussioncomment-9473072:

> We might not even need to write too much new code for this, I suppose. Given that models are separate, we can start (main_A + speculative) on instance_A, (main_B + speculative) on instance_B. Then we need to orchestrate the data/logic passing during transition phase:
> 
> In the 'middle' of main model processing (A is done with first half), we need to pass activations to B and whatever B speculated so far back to A
> At the end of main model processing (B is done with logits) we need to get whatever latest speculation on B is, consolidate it with what we have currently produced on A, pass the 'current approved tokens' to A, start speculating on B.
> repeat

Relevant links:
* https://github.com/ggerganov/llama.cpp/discussions/6853
* https://github.com/ggerganov/llama.cpp/pull/3624
* https://github.com/ggerganov/llama.cpp/pull/6829

Devices I can test it on are:
* M2 Ultra 192GB
* M2 24GB
* M1 16GB

**Comment by okuvshynov** [comment]
Work will be done here: https://github.com/okuvshynov/llama.cpp/tree/duo, as I'll need some changes to llama.cpp itself.
